publications/radstack/radstack.tex

```diff
@@ -0,0 +1,1148 @@
+was too slow to power an interactive application. We quickly realized a row
+store, and relational model as a whole, was a sub-optimal choice for the
+application we were trying to build.
+
+\begin{figure}
+\centering
+\includegraphics[width = 2.6in]{precompute}
+\caption{Precomputating queries scales exponentially. To support every possible
+combination of queries, even a small data set can generate many possible
+queries.
+}
+\label{fig:precompute}
+\end{figure}
+
+\begin{figure}
+\centering
+\includegraphics[width = 2.6in]{radstack}
+\caption{
+The components of the RADStack. Kafka acts as the event delivery
+endpoints. Samza and Hadoop process data to load data into Druid. Druid acts as
+the endpoint for queries.
+}
+\label{fig:radstack}
+\end{figure}
+
+We next evaluated HBase \cite{george2011hbase}, a NoSQL key/value store. As is
+common with using key/value stores, we precomputed the total set of queries we
+anticipated users would make. An example of this precomputation is shown in
+Figure~\ref{fig:precompute}. Our results with HBase are shown in
+Table~\ref{tab:hbase_results} for 500, 000 records. 
+Queries were acceptably fast with this solution as we were effectively doing
+O(1) lookups into maps.  However, the solution was not particularly flexible;
+if something wasn't precomputed, it wasn’t queryable, and the precomputation
+time became a big pain point. We attempted to build iceberg cubes as described
+in \cite{beyer1999bottom}, restricting the space of our pre-computations, but
+even then the precomputation times were unacceptable.  Our entire evaluation of
+both RDBMS and key/value stores is described in detail in
+\cite{tschetter2011druid}.
+
+Druid was developed in an attempt to solve some of the problems seen with
+traditional serving layers. Druid was designed from the ground up to provide
+arbitrary data exploration, low latency aggregations, and fast data ingestion.
+Druid was also designed to intake fully denormalized data, and moves away from
+the traditional relational model. Since much of raw data is not denormalized,
+it must be processed before it can be ingested and queried. Multiple streams of
+data had to be joined, cleaned up, and transformed before it was usable in
+Druid, but that was the trade-off in order to get the performance necessary to
+power an interactive data application. We introduced stream processing to our
+stack to provide the processing required before raw data could be loaded into
+Druid. Our streaming processing jobs range from simple data transformations,
+such as id to name lookups, up to complex operations such as multi-stream
+joins. Pairing Druid with a stream processor enabled flexible data processing
+and querying, but because events were delivered from many different locations
+and sources, we also required a message bus for high throughput event delivery.
+Kafka was added to our stack, and Kafka became the event delivery endpoint for
```

Comment:
Honestly, everything from the introduction of Kafka on seems a little rushed.  Not sure if that's due to space considerations, but it would probably be nice to flesh it out a bit more.